### ISSUE TITLE
Lab 7 Level 5

### ISSUE THREAD

**Issue by isaactyj** [nus-cs2030/2021-s2#179]
# Summary

I am unable to get f to recursively call itself. Can someone help me please:/

## Description:

This is the error that I get

 attempted to call method f(int) which cannot be invoked until method flatmap((x)->f(x)), and method flatmap((x)->f(3 * x + 1)) are declared

## Code Snippets

Logger<Integer> f(int n) {
    if (n == 1) {
        return Logger.make(1);
    } else if (n % 2 == 0) {
        return Logger.make(n).flatmap(x -> f(x));
    } else {
        return Logger.make(n).flatmap(x -> f(3 * x + 1));
    }
}
## Screenshots

Screenshots (if any)

**Comment by seancxy** [comment]
the M in flatMap isn't capitalized!

**Comment by isaactyj** [comment]
@seanlowjk dont think its the main issue, even after changing it, the f function does not work

**Comment by ShyamalSankar** [comment]
Note that f takes in an integer. Maybe using a loop would be easier?

**Comment by isaactyj** [comment]
@ShyamalSankar care to elaborate wdym?

**Comment by ShyamalSankar** [comment]
you can consider iteratively getting the next logger element using a for or a while loop to get the next Logger until the the logger contains 1.

**Comment by isaactyj** [comment]
Logger(int n) {
     while (n!=1) {
          if(n%2==0) {
                  return Logger.make(n).flatmap(x -> Logger.make(x).map(y -> f(y/2)))
           else { 
                  return  Logger.make(n).flatmap(x -> Logger.make(x).map(y -> f(3*y/ + 1)))
           }
       }
}

Do u mean like this?

**Comment by ShyamalSankar** [comment]
Something like that but you can initialise the variable with Logger.make(n) and then keep updating it with each iteration. Then at the end you return the product

**Comment by kexinnnwang** [comment]
Hi I did something similar to your original answer. I think you are missing one part on updating the value of Logger in each step. In your recursive step, you are creating a new Logger using either (y / 2) or (3 * y + 1), so you should include a step to update the value before calling f.

**Comment by Keithlimhs** [comment]
Hi, you can implement this function recursively using a helper function. Basically the helper function takes in a Logger instance and does the same as f. Then in the original f(int n), you can call the helper function after creating Logger.make(n). Hope this helps

**Comment by isaactyj** [comment]
@ShyamalSankar @Keithlimhs @kexinnnwang 

Logger<Integer> f(int n) {
    if (n == 1) {
        return Logger.make(1);
    } else {
        Logger<Integer> a = Logger.make(n);
        while(a.getObj() != 1) {
            if(a.getObj() % 2 == 0){
                a.map(x -> x / 2);
            } else {
                a.map(x -> 3 * x + 1);
            }
        }
        return a;
    }
}

I did this can you help me point out where is the error?

**Comment by Keithlimhs** [comment]
Hi, if i'm not wrong, the output will only be the initialized String, because the map() method will return a new LoggerImpl whenever it is called, so when you return a, you are essentially returning the LoggerImpl that is created when you initialized a

**Comment by ShyamalSankar** [comment]
Yes keith is right because map returns a new object instead of mutating the original so you should reassign the variable. Ie 
a = a.map.(...

**Comment by kexinnnwang** [comment]
Yup this only creates a Logger using the given value. You can try to combine this answer with your original answer to include calling f recursively.

**Comment by dumbasssoul** [comment]
hello i did it like this
`Logger<Integer> f(int n) {
    Logger<Integer> log = Logger.make(n);
    while (n != 1) {
        if (n % 2 == 0) {
            log = log.map(x -> x / 2);
            n /= 2;
        } else if (n % 2 == 0) {
            log = log.map(x -> 3 * x + 1);
            n = 3 * n + 1;
        }
    }
    return log;
}`

this works for f(16) but not for f(10) as it takes too long to run(?) is there anything wrong with this?

**Comment by radellng** [comment]
`Logger<Integer> f(int n) {
    if (n == 1) {
        return Logger.make(1);
    } 
    Logger<integer> a = Logger.make(n);
    if (n % 2 == 0) {
        a.map(x -> x / 2);
        n /= 2;
    } else {
        a.map(x -> 3 * x + 1);
        n = 3 * n + 1;
    }
    return a;
}`

Hello i did this and i got this error. Please advice, or is my method wrong.
![forum](https://user-images.githubusercontent.com/77708166/113902895-f99b6780-9802-11eb-9a6e-e5e8ad33c944.png)

**Comment by adellelimrj** [comment]
> Yes keith is right because map returns a new object instead of mutating the original so you should reassign the variable. Ie
> a = a.map.(...

in addition to this, i think you should drop the if statement, because that will cause the end to return Logger.make(1) which is not desirable as it does not have any of the map() methods implemented

**Comment by AdelineWongg** [comment]
> @ShyamalSankar @Keithlimhs @kexinnnwang
> ![function](https://user-images.githubusercontent.com/77622789/113756941-9ba84a80-9744-11eb-8922-9c9f86f2c21b.JPG)
> 
> I did this can you help me point out where is the error?

I think the return a in the 3rd last line is correct but the problem here is that his a is not being updated.
It should have been a = a.map(insert the respective manipulation)

**Comment by IntrovertHedgehog** [comment]
Maybe there's a problem with flatMap() ?

**Comment by hendricksgin** [comment]
> `Logger<Integer> f(int n) { if (n == 1) { return Logger.make(1); } Logger<integer> a = Logger.make(n); if (n % 2 == 0) { a.map(x -> x / 2); n /= 2; } else { a.map(x -> 3 * x + 1); n = 3 * n + 1; } return a; }`
> 
> Hello i did this and i got this error. Please advice, or is my method wrong.
> ![forum](https://user-images.githubusercontent.com/77708166/113902895-f99b6780-9802-11eb-9a6e-e5e8ad33c944.png)

your I for Integer should be capitalised (for "Logger<integer> a = Logger.make(n)")

**Comment by random689** [comment]
> > @ShyamalSankar @Keithlimhs @kexinnnwang
> > ![function](https://user-images.githubusercontent.com/77622789/113756941-9ba84a80-9744-11eb-8922-9c9f86f2c21b.JPG)
> > I did this can you help me point out where is the error?
> 
> I think the return a in the 3rd last line is correct but the problem here is that his a is not being updated.
> It should have been a = a.map(insert the respective manipulation)
> also for the else part,
> based on the example(f(10)) given on codecrunch, X_3 and then +1 to the result should be done separately a = a.map(x -> x_3) // this gives 15
> a = a.map(x -> x +1) // this gives 16

I think you should also use a.test(x -> x != 1) instead since they asked you to make the test function as well.

**Comment by Deshun99** [comment]
![Untitled](https://user-images.githubusercontent.com/77624433/114036055-5013af00-98b2-11eb-9962-dbbf866ffc17.png)

i tried doing this implementation, if the number is odd map the multiplication and the addition part separately

**Comment by yokester98** [comment]
I did this, and it resulted in a reverse order lol. Not sure how to flip it around though? 

`Logger<Integer> f(int n) {
   if (n == 1) {
       return Logger.make(1);
   } else if (n % 2 == 0) {
       return f(n / 2).flatMap(x -> Logger.make(x).map(y -> y * 2));
   } else {
       return f((3 * n) + 1).flatMap(x -> Logger.make(x).map(y -> y - 1)
        .flatMap(y -> Logger.make(y).map(z -> z / 3)));
   }
}`

(Nicely indented code)
![image](https://user-images.githubusercontent.com/39672956/114057848-86a6f500-98c5-11eb-80d8-11424aa48940.png)


(Output)
![image](https://user-images.githubusercontent.com/39672956/114057579-421b5980-98c5-11eb-8c7d-43b708d1566d.png)

**Comment by yokester98** [comment]
Hello @dumbasssoul @radellng, heres a suggestion! The code below is incomplete but you can get the idea!

`Logger<Integer> f(int n) {
    return helper(Logger.make(n)); 
}`

`Logger<Integer> helper(Logger<Integer> logger) {
    if (logger.test(x -> x == 1)) {
        return logger;
    } else if (logger.test(x -> x % 2 == 0)) {
        return helper(logger.map(x -> x / 2));
    } else {
        // insert your code here
    } 
}`

(nicely indented code)
![image](https://user-images.githubusercontent.com/39672956/114061790-6711cb80-98c9-11eb-88d9-968df32f63c6.png)

**Comment by yokester98** [comment]
> erm not quite. You do not have to recursively call f or use a flatMap.
> 
> you should initialise a logger containing n and then with each iteration of the while loop update the variable
> 
> So
> logger object initialised with the input int
> while (logger content not equal to 1)
> if logger even -> map the logger variable to divide its content by 2
> if logger odd -> multiply by 3 and add 1
> once while loop ends you can return the product because the logger would definitely contain 1 along with the logs of all the values it undertook.

Hi @ShyamalSankar, I think this question is asking for a recursive solution. You can check out my approach above!

**Comment by radellng** [comment]
@yokester98 Thank you for your insight! Cheers friend!

**Comment by espoirsetreves** [comment]
Hello! Please check out the Java API, think it'll help you slightly.

https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/stream/Stream.html#flatMap(java.util.function.Function)
